File: Java/labs/labs12_Flyweight/task_2_1 – solution/src/VehicleFactory.java
import java.util.HashMap;

public class VehicleFactory {
    HashMap<Integer, Vehicle> vehicles = new HashMap<>();

    public Vehicle getVehicle(int type){
        Vehicle vehicle = vehicles.get(type);
        if (vehicle == null){
            if (type == 0){
                vehicle = new Car();
            } else if (type == 1){
                vehicle = new Truck();
            }
            vehicles.put(type, vehicle);
        }
        return vehicle;
    }
}

File: Cpp/lab12_Flyweight/task_2_1 - solution/src/vehicle_factory.cpp
#include "vehicle_factory.hpp"


Vehicle* VehicleFactory::getVehicle(int type)
{
  auto iter = this->vehicles.find(type);

  if (iter == this->vehicles.end())
  {
    Vehicle* vehicle;
    switch (type)
    {
      case 0:
      {
        vehicle = new Car();
        break;
      }
      case 1:
      {
        vehicle = new Truck();
        break;
      }
    }

    iter = this->vehicles.emplace_hint(iter, type, vehicle);
  }

  return std::get<1>(*iter);
}
